

##### Raw data processing

1, convert BCL files to fastq format by bcl2fastq

```shell
#note, put the SampleSheet.csv file under 210920_NB501557_0142_AHTGHFBGXG folder (NextSeq result folder)
bcl2fastq --runfolder-dir 210920_NB501557_0142_AHTGHFBGXG/ -p 64 --output-dir 210920_NB501557_0142_AHTGHFBGXG/fastq --no-lane-splitting --minimum-trimmed-read-length 5 --mask-short-adapter-reads 5
```

2, trim reads by cutadapt.

```shell
#For data generated by the NextSeq, use '--nextseq-trim=20' . 
#For read1 which contains mRNA conding information, keep at least 22nt for downstream reads mapping. 
#For read2, the first 8 nts are UMI and the following 6 nts are barcodes. 
#Perform FastQC (optional).

cutadapt --cores=64 --nextseq-trim=20 -m 22:14 -o Qlastsp_R1.fastq.gz -p Qlastsp_R2.fastq.gz lastsp_S1_R1_001.fastq.gz lastsp_S1_R2_001.fastq.gz

```

3, run zUMIs pipeline (follow zUMIs mannual scripts).

```shell
#clone the zUMI repository
git clone https://github.com/sdparekh/zUMIs.git
```

```shell
#prepare genome reference
#merge genome reference and ERCC reference or import the ERCC reference by zUMIs config file
#here, we append the ERCC file to genome file.
cat ERCC92.fa >> GRCh38.p13.genome.fa
cat ERCC92.gtf >> gencode.v33.annotation.gtf
```

```shell
#build STAR index.
#It is not necessary to generate the genome index with specific overhang and splice-site reference.
#zUMIs passes the GTF file to STAR while mapping to insert junctions on the fly.
STAR --runMode genomeGenerate --runThreadN 64 --genomeDir /data/lyuj2/STAR_ref/index/STAR2.7.3a/ --limitGenomeGenerateRAM 111000000000 --genomeFastaFiles /data/lyuj2/STAR_ref/index/STAR/GRCH38_v33/GRCh38.p13.genomeE.fa
```

```
configure YAML config file by online Shiny app or use zUMIs template file. 
note, the zUMIs can generate demultipled bam file,yet it's not compatible with UMI-tools
```

```shell
#run zUMIs
#-c flag to use conda
#-y flag to use yaml file (see the examplory yaml file for LAST-seq)
/data/lyuj2/zUMIs/zUMIs.sh -c -y lastsp.yaml
```

##### Count matrix processing

zUMIs produces dge output in .rds format that can be read in R with the following command.

```R
AllCounts <- readRDS("zUMIs_output/expression/example.dgecounts.rds")
names(AllCounts)
[1] "umicount"  "readcount"

names(AllCounts$umicount)
[1] "exon"   "inex"   "intron"

names(AllCounts$umicount$exon)
[1] "all"          "downsampling"
```

```
refer to zUMIs mannual:
AllCounts is a list of lists with all the count matrices as sparseMatrix. The parent list contains UMI and read count quantification. In each of these counting types, you will find the three feature types (introns,exons and intron+exon). Each of those contain a sparseMatrix generating using all reads observed and a list for the downsampling sizes requested. Each of the downsampling list elements is also a sparseMatrix.

The sparseMatrix can be converted into a conventional count table using "as.matrix" function in R and saved as a text file using the code below.
```

```shell
dge <- as.matrix(AllCounts$umicount$exon$all)
#note: It is highly recommended to retain sparseMatrix format for the downstream analysis, especially when you have >20,000 cells to save time and space.
```



