---
title: "Fig3"
author: "Lyu"
date: "7/19/2022"
output: html_document
---

```{r}
library(tidyverse)
```

```{r define function to calculate SEM}
#+++++++++++++++++++++++++
# Function to calculate the mean and the standard error of mean
# for each group
#+++++++++++++++++++++++++
# data : a data frame
# varname : the name of a column containing the variable
#to be summariezed
# groupnames : vector of column names to be used as
# grouping variables
data_summary <- function(data, varname, groupnames){
  require(plyr)
  summary_func <- function(x, col){
    c(mean = mean(x[[col]], na.rm=TRUE),
      SEM = sd(x[[col]], na.rm=TRUE)/sqrt(length(x[[col]])))
  }
  data_sum<-ddply(data, groupnames, .fun=summary_func,
                  varname)
  data_sum <- rename(data_sum, c("mean" = varname))
  return(data_sum)
}

```

```{r technical noise Fig3a}
LASTCV <- readRDS('averageCV_mean_1.rds')
CELCV <- readRDS('averageCV_mean_cel.rds')
SMARTCV <- readRDS('averageCV_mean_smart.rds')

LASTCV$ID <- rownames(LASTCV)
LASTCV <-  LASTCV%>%
  filter(averagemean>0.5)

CELCV$ID <- rownames(CELCV)
CELCV <-  CELCV%>%
  filter(averagemean>0.5)

SMARTCV$ID <- rownames(SMARTCV)

LAST_vs_CEL <- inner_join(LASTCV,CELCV,by='ID')

LAST_vs_CEL_vs_SMART <- inner_join(LAST_vs_CEL,SMARTCV)

colnames(LAST_vs_CEL) <- c('lastCV','lastMean','ID','celCV','celMean')

LAST_vs_CEL_vs_SMARTLong <- LAST_vs_CEL_vs_SMART%>%
  pivot_longer(cols = c('lastCV','celCV','smartCV'),names_to ='names', values_to='values')
LAST_vs_CEL_vs_SMARTLong$names <- factor(LAST_vs_CEL_vs_SMARTLong$names,levels = c('lastCV','celCV','smartCV'))

LAST_vs_CEL_vs_SMARTLong <- LAST_vs_CEL_vs_SMARTLong %>% group_by(names)


p <-  ggplot(LAST_vs_CEL_vs_SMARTLong,aes(x=values,color=names))+
  geom_density(size=0.25)+
  theme_classic() +
  theme(
    axis.text.y = element_text(size=6),
    axis.text.x = element_text(size = 6),
    axis.title = element_text(colour='black', size=7),
    legend.title=element_blank(),
    #legend.key=element_blank(),
    axis.line.x = element_line(colour = "black", size=0.25),
    axis.line.y = element_line(colour = "black", size=0.25),
    axis.ticks.x = element_line(colour = "black", size = 0.25),
    axis.ticks.y = element_line(colour = "black", size = 0.25),
    legend.position = c(0.8, 0.8),
         legend.key.size = unit(0.1, "cm"),
         legend.key.width = unit(0.1, "cm"),
         legend.text = element_text(size=6))+
   #xlim(-0.2,0.2)+
   ylab('density')+
   xlab("technical noise (CV)")
   #annotate(geom="text",x=0.2, y=4, label="WT",size=2)
 
 ggsave(filename = "technical noise_all.pdf",plot = p,width=90,height = 45,units = "mm", dpi = 300)
```

```{r gene detection SMART-seq Fig3b}
LAST_vs_SMART <- read.csv("./input/gene_LAST_SMART_1M.csv",sep = ",",header = T)
LAST <- LAST_vs_SMART[1:10,]
SMART <- LAST_vs_SMART[21:29,]
LAST <- LAST %>%
  pivot_longer(cols = X50K:X1000K,names_to='name',values_to='value')

SMART <- SMART %>%
  pivot_longer(cols = X50K:X1000K,names_to='name',values_to='value')
#####generate error bar
errorbarLAST <- data_summary(LAST,varname='value',groupnames='name')
errorbarLAST$type <- rep('LAST',5)
errorbarSMART <- data_summary(SMART,varname='value',groupnames='name')
errorbarSMART$type <- rep('SMART',5)

errorbar <- rbind(errorbarLAST,errorbarSMART)
errorbar$read <- c('1M','0.1M','0.2M','0.5M','0.05M','1M','0.1M','0.2M','0.5M','0.05M')

p <- ggplot(data=errorbar, aes(x=read, y=value,fill=type)) +
  geom_bar(stat="identity",width=0.8,position=position_dodge())+
  geom_errorbar(aes(ymin=value-SEM, ymax=value+SEM),width=.25,size=0.25,position=position_dodge(.9))+
  theme_classic() +
  theme(
    axis.text.y = element_text(size=6),
    axis.text.x = element_text(size = 6),
    axis.title = element_text(colour='black', size=7),
    legend.title=element_blank(),
    #legend.key=element_blank(),
    axis.line.x = element_line(colour = "black", size=0.25),
    axis.line.y = element_line(colour = "black", size=0.25),
    axis.ticks.x = element_line(colour = "black", size = 0.25),
    axis.ticks.y = element_line(colour = "black", size = 0.25),
    strip.background=element_blank(),
    strip.text=element_text(size=7),
    legend.position = c(0.25, 0.9),
    legend.key.size = unit(0.05, "cm"),
    legend.key.width = unit(0.05, "cm"),
    legend.text = element_text(size=4))+
  xlab("reads#")+
  ylab("gene#")+
  scale_fill_discrete(labels=c("LAST-seq","SMART-seq"))
  
ggsave(filename = "LAST_SMART.pdf",plot = p,width=45,height = 45,units = "mm", dpi = 300)
```

```{r UMI/overlapping genes Fig3c}
errorbar_gene_UMIss <- readRDS('./input/errorbar_gene_UMIss.rds')

p <- ggplot(data=errorbar_gene_UMIss[c(4,8),], aes(x=type, y=value,fill=type)) +
  geom_bar(stat="identity",width=0.8)+
  geom_errorbar(aes(ymin=value-SEM, ymax=value+SEM),width=.25,size=0.25)+
  theme_classic() +
  theme(
    axis.text.y = element_text(size=6),
    axis.text.x=element_blank(),
    axis.title = element_text(colour='black', size=7),
    legend.title=element_blank(),
    #legend.key=element_blank(),
    axis.line.x = element_line(colour = "black", size=0.25),
    axis.line.y = element_line(colour = "black", size=0.25),
    axis.ticks.x=element_blank(),
    axis.ticks.y = element_line(colour = "black", size = 0.25),
    strip.background=element_blank(),
    strip.text=element_text(size=7),
    legend.position = c(0.9, 1),
    legend.key.size = unit(0.1, "cm"),
    legend.key.width = unit(0.1, "cm"),
    legend.text = element_text(size=5))+
  scale_y_continuous(breaks=seq(0,30,5))+
  scale_fill_discrete(labels = c("LASTseq", "CELseq2"))+
  xlab("")+
  ylab("UMI#/overlapping gene")

ggsave(filename = "LAST_CELss_UMIpergene_overlapping.pdf",plot = p,width=45,height = 45,units = "mm", dpi = 300)
```


```{r gene detection by CEL-seq2 Fig3d}
celss <- readRDS('./input/celss_1M.rds')

cellssrm <- celss[!grepl('^ERCC-',rownames(celss)),]

cellssrmf <-na.omit(as.data.frame(cellssrm[rowSums(cellssrm>0)>3,])) 

#count genes 
celss_genef <- colSums(cellssrmf != 0)

cellssrmff <- cellssrm[rowSums(cellssrm>0)>9,]
#count genes 
celss_geneff <- colSums(cellssrmff != 0)


celgene <- as.data.frame(cbind(celss_gene,celss_genef,celss_geneff))
colnames(celgene)=c('f1','f4','f10')

celgene <- celgene %>%
  pivot_longer(cols = f1:f10 ,names_to="name",values_to='value')

#######last filter
last_UMI_1M_1A <-as.data.frame(readRDS('./input/last_UMI_1M_1.rds'))
last_UMI_1M_1 <- last_UMI_1M_1A[!grepl('^ERCC-',rownames(last_UMI_1M_1A)),]
last <- as.data.frame(last_UMI_1)

lastf <-na.omit(last[rowSums(last>0)>3,]) 
#count genes 
last_genef <- colSums(lastf != 0, na.rm = T)

lastff <-na.omit(last[rowSums(last>0)>9,]) 
#count genes 
last_geneff <- colSums(lastff != 0, na.rm = T)

lastgene <- as.data.frame(cbind(last_gene,last_genef,last_geneff))
colnames(lastgene)=c('f1','f4','f10')


lastgene <- lastgene %>%
  pivot_longer(cols = f1:f10,names_to="name",values_to='value')

#####generate error bar
errorbarLASTgene <- data_summary(lastgene,varname='value',groupnames='name')
errorbarLASTgene$type <- rep('LAST',3)
errorbarCELssgene <- data_summary(celgene,varname='value',groupnames='name')
errorbarCELssgene$type <- rep('CEL',3)

errorbargene <- rbind(errorbarLASTgene,errorbarCELssgene)
errorbargene$type <- factor(errorbargene$type,levels = c('LAST','CEL'))
errorbargene$name <- factor(errorbargene$name,levels = c('f1','f4','f10'))

p <- ggplot(data=errorbargene, aes(x=type, y=value,fill=name)) +
  geom_bar(stat="identity",width=0.8,position="identity",alpha=.3)+
  geom_errorbar(aes(ymin=value-SEM, ymax=value+SEM),width=.25,size=0.25,position=position_dodge(.9))+
  theme_classic() +
  theme(
    axis.text.y = element_text(size=6),
    axis.text.x=element_text(size=6),
    axis.title = element_text(colour='black', size=7),
    legend.title=element_blank(),
    #legend.key=element_blank(),
    axis.line.x = element_line(colour = "black", size=0.25),
    axis.line.y = element_line(colour = "black", size=0.25),
    axis.ticks.x=element_blank(),
    axis.ticks.y = element_line(colour = "black", size = 0.25),
    strip.background=element_blank(),
    strip.text=element_text(size=7),
    legend.position = c(0.2, 1),
    legend.key.size = unit(0.1, "cm"),
    legend.key.width = unit(0.1, "cm"),
    legend.text = element_text(size=5))+
  scale_x_discrete(breaks=c("LAST","CEL"),
                   labels=c("LASTseq", "CELseq2"))+
  xlab("")+
  ylab("gene#")

ggsave(filename = "LAST_CELss_genefilter.pdf",plot = p,width=45,height = 45,units = "mm", dpi = 300)
```

```{r CEL UMI distribution of gene Fig3e follows Fig3d}
##########celseq, the UMI of genes detected in 10 cells vs < 10 cells

#genes detected in more than 3 cells
cellssrmff <-na.omit(as.data.frame(cellssrm[rowSums(cellssrm>0)>9,])) 
#genes detected in less than 4 cells
Ocellssrmff <-na.omit(as.data.frame(cellssrm[!rowSums(cellssrm>0)>9,])) 

colSums(cellssrmff>0)
colSums(Ocellssrmff>0)

cell10g <- cellssrmff[,1]
cell10g <-as.data.frame(cell10g[cell10g>0])
colnames(cell10g) <- 'UMI'
cell10g$type <- rep('genes>9',6668)

Ocell10g <- Ocellssrmff[,1]
Ocell10g <-as.data.frame(Ocell10g[Ocell10g>0]) 
colnames(Ocell10g) <- 'UMI'
Ocell10g$type <- rep('genes<=9',7098)

cell10gmeg <- rbind(cell10g,Ocell10g)

library(tidyverse)
library(ggpubr)
library(scales)
my_comparisons <- list(c('genes>9','genes<=9'))
p <- cell10gmeg %>% ggplot(aes(x=type,y=UMI,color=type))+
  geom_boxplot(lwd=0.25, width=0.75, fatten=1,outlier.size = 0.1)+
  scale_y_log10(labels = number_format(accuracy = 1))+
  stat_compare_means(comparisons = my_comparisons,size=1.5)+
  theme_man+
  #change x and y labels
  xlab(" ")+  
  ylab("UMI#")+
  theme(legend.position='none')

ggsave(filename = "cel_gene_in_10cells.pdf",plot = p,width=45,height = 45,units = "mm", dpi = 300)

```

```{r LAST UMI distribution of gene Fig3d follows Fig3d}
##########lastseq, the UMI of genes detected >= 4 cells vs < 4 cells

#genes detected in more than 3 cells
lastff <-na.omit(last[rowSums(last>0)>9,]) 
#genes detected in less than 4 cells
Olastff <-na.omit(last[!rowSums(last>0)>9,]) 

colSums(lastff>0)
colSums(Olastff>0)

last10g <- lastff[,1]
last10g <-as.data.frame(last10g[last10g>0])
colnames(last10g) <- 'UMI'
last10g$type <- rep('genes>9',6233)

Olast10g <- Olastff[,1]
Olast10g <-as.data.frame(Olast10g[Olast10g>0]) 
colnames(Olast10g) <- 'UMI'
Olast10g$type <- rep('genes<=9',4799)

last10gmeg <- rbind(last10g,Olast10g)

library(tidyverse)
library(ggpubr)
library(scales)
my_comparisons <- list(c('genes>9','genes<=9'))
p <- last10gmeg %>% ggplot(aes(x=type,y=UMI,color=type))+
  geom_boxplot(lwd=0.25, width=0.75, fatten=1,outlier.size = 0.1)+
  scale_y_log10(labels = number_format(accuracy = 1))+
  stat_compare_means(comparisons = my_comparisons,size=1.5)+
  theme_man+
  #change x and y labels
  xlab(" ")+  
  ylab("UMI#")+
  theme(legend.position='none')

ggsave(filename = "last_gene_in_10cells.pdf",plot = p,width=45,height = 45,units = "mm", dpi = 300)
```

```{r mapping rate comparison LASTseq vs CEL-seq Fig3f}
errorbarLAST_exon_intron <- readRDS('./input/errorbarlastmapping.rds')
errorbarLAST_exon_intron$type <- rep('LAST',5)

CEL_exon_intron <- read.delim("./input/CELSS.readspercell.txt",header = T)
CEL_exon_intron <- CEL_exon_intron %>%
  filter(RG  != 'bad')

CEL_exon_intron <- CEL_exon_intron %>% pivot_wider(names_from = RG, values_from = N)
CEL_exon_intron <-as.data.frame(CEL_exon_intron)
rownames(CEL_exon_intron) <- CEL_exon_intron$type
CEL_exon_intron$type <- NULL
CEL_exon_intron <-as.data.frame(t(CEL_exon_intron)) 

CEL_exon_intron <- CEL_exon_intron %>%
  mutate(exonmapping = 100*(Exon/rowSums(CEL_exon_intron)))
CEL_exon_intron <- CEL_exon_intron %>%
  mutate(intronmapping = 100*(Intron/rowSums(CEL_exon_intron)))
CEL_exon_intron <- CEL_exon_intron %>%
  mutate(intergenicmapping = 100*(Intergenic/rowSums(CEL_exon_intron)))
CEL_exon_intron <- CEL_exon_intron %>%
  mutate(totalmapping = 100*(Ambiguity/rowSums(CEL_exon_intron)))
CEL_exon_intron <- CEL_exon_intron %>%
  mutate(mappedrate = 100*(1-Unmapped/rowSums(CEL_exon_intron)))

CEL_exon_intronrate <- CEL_exon_intron[6:10]
colnames(CEL_exon_intronrate) <- c("Exon","Intron","Intergenic","Ambiguity","Total")

rowSums(CEL_exon_intron[1:5])


##convert to long
CEL_exon_intronrate <- CEL_exon_intronrate %>%
  pivot_longer(cols=Exon:Total, names_to = 'Type',values_to='Rate')

errorbarCEL_exon_intronrate <- data_summary(CEL_exon_intronrate,varname='Rate',groupnames='Type')
errorbarCEL_exon_intronrate$type <- rep('CEL',5)

errorbar_exon_intron <- rbind(errorbarLAST_exon_intron,errorbarCEL_exon_intronrate)
errorbar_exon_intron$type <- factor(errorbar_exon_intron$type,levels = c('LAST','CEL'))
errorbar_exon_intron$Type <- factor(errorbar_exon_intron$Type,levels = c('Total','Exon','Intron',"Intergenic",'Ambiguity'))

p <- ggplot(data=errorbar_exon_intron, aes(x=Type, y=Rate,fill=type)) +
  geom_bar(stat="identity",width=0.8,position=position_dodge())+
  geom_errorbar(aes(ymin=Rate-SEM, ymax=Rate+SEM),width=.25,size=0.25,position=position_dodge(.9))+
  theme_classic() +
  theme(
    axis.text.y = element_text(size=6),
    axis.text.x = element_text(size = 4),
    axis.title = element_text(colour='black', size=7),
    legend.title=element_blank(),
    #legend.key=element_blank(),
    axis.line.x = element_line(colour = "black", size=0.25),
    axis.line.y = element_line(colour = "black", size=0.25),
    axis.ticks.x = element_line(colour = "black", size = 0.25),
    axis.ticks.y = element_line(colour = "black", size = 0.25),
    strip.background=element_blank(),
    strip.text=element_text(size=7),
    legend.position = c(0.9, 0.9),
    legend.key.size = unit(0.1, "cm"),
    legend.key.width = unit(0.1, "cm"),
    legend.text = element_text(size=5))+
  scale_fill_discrete(labels = c("LASTseq", "CELseq2"))+
  xlab(" ")+
  ylab("Mapping Rate(%)")

ggsave(filename = "LAST_CELss_mapping_all.pdf",plot = p,width=45,height = 45,units = "mm", dpi = 300)

```


